### PR TITLE
add CRUD poster

### PR DESCRIPTION
**Pull request by yaroslavvasilenko**

### CODE REVIEW

File: database/migrate.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	_ "github.com/jackc/pgx/v5/stdlib"
+	"github.com/pkg/errors"
+	"github.com/pressly/goose/v3"
+)
+
+//go:embed migrations/*.sql
+var migrations embed.FS
+
+func Migrate(url string) error {
+	db, err := sql.Open("pgx", url)
```
Comment (cringoleg) [inline database/migrate.go:18]: Why is it pgx exactly?
Maybe pass some variable here?

File: database/migrate.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	if err := goose.SetDialect("postgres"); err != nil {
+		return err
+	}
+
+	version, err := goose.GetDBVersion(db)
+	if err != nil {
+		return err
+	}
+
+	err = goose.Up(db, "migrations")
```
Comment (cringoleg) [inline database/migrate.go]: Move "migrations" to const

File: database/migrate.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+	if err != nil {
+		return errors.Wrap(err, "cannot connect to db")
+	}
+	defer db.Close()
+
+	if err = db.Ping(); err != nil {
+		return errors.Wrap(err, "cannot ping db")
+	}
+	goose.SetBaseFS(migrations)
+	if err := goose.SetDialect("postgres"); err != nil {
```
Comment (cringoleg) [inline database/migrate.go]: Move "postgres" to const

File: database/migrate.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	}
+
+	version, err := goose.GetDBVersion(db)
+	if err != nil {
+		return err
+	}
+
+	err = goose.Up(db, "migrations")
+	if err != nil {
+		if err := goose.DownTo(db, "migrations", version); err != nil {
```
Comment (cringoleg) [inline database/migrate.go]: Move "migrations" to const

File: internal/storage.go [hunk 5]
```diff
@@ -1,11 +1,57 @@
...
+		return err
+	}
+	return nil
+}
+
+func (s *Storage) GetPoster(pID string) (entity.Poster, error) {
+	poster := entity.Poster{}
+
+	err := s.gorm.Table(posterTable).First(&poster).Where("id = ?, deleted_at IS NULL", pID).
+		Error
```
Comment (cringoleg) [inline internal/storage.go:35]: Looks ugly

File: internal/service.go [hunk 6]
```diff
@@ -11,3 +17,45 @@ func NewService(s *Storage) *Service {
...
+		return entity.Poster{}, err
+	}
+
+	return poster, nil
+}
+
+func (s *Service) DeletePoster(pID string) error {
+	err := s.s.DeletePoster(pID)
+	if err != nil {
+		return err
```
Comment (cringoleg) [inline internal/service.go:46]: `return errors.Wrap(err, "Deleting poster")`

File: internal/service.go [hunk 7]
```diff
@@ -11,3 +17,45 @@ func NewService(s *Storage) *Service {
...
+
+	return nil
+}
+
+func (s *Service) UpdatePoster(p entity.Poster) error {
+	timeNow := time.Now()
+	p.UpdatedAt = timeNow
+	err := s.s.UpdatePoster(p)
+	if err != nil {
+		return err
```
Comment (cringoleg) [inline internal/service.go:57]: `return errors.Wrap(err, "Updating poster")`

File: database/migrate.go [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+func Migrate(url string) error {
+	db, err := sql.Open("pgx", url)
+	if err != nil {
+		return errors.Wrap(err, "cannot connect to db")
+	}
+	defer db.Close()
+
+	if err = db.Ping(); err != nil {
+		return errors.Wrap(err, "cannot ping db")
```
Comment (cringoleg) [inline database/migrate.go]: `return errors.Wrap(err, "pinging db")`

File: database/migrate.go [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+	"github.com/pressly/goose/v3"
+)
+
+//go:embed migrations/*.sql
+var migrations embed.FS
+
+func Migrate(url string) error {
+	db, err := sql.Open("pgx", url)
+	if err != nil {
+		return errors.Wrap(err, "cannot connect to db")
```
Comment (cringoleg) [inline database/migrate.go]: `return errors.Wrap(err, "connecting to db")`

File: database/migrate.go [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+		return errors.Wrap(err, "cannot ping db")
+	}
+	goose.SetBaseFS(migrations)
+	if err := goose.SetDialect("postgres"); err != nil {
+		return err
+	}
+
+	version, err := goose.GetDBVersion(db)
+	if err != nil {
+		return err
```
Comment (cringoleg) [inline database/migrate.go]: `return errors.Wrap(err, "getting db version")`

File: database/migrate.go [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+	}
+
+	err = goose.Up(db, "migrations")
+	if err != nil {
+		if err := goose.DownTo(db, "migrations", version); err != nil {
+			return err
+		}
+
+		return err
```
Comment (cringoleg) [inline database/migrate.go]: Wrap these errors too

File: internal/controller.go [hunk 12]
```diff
@@ -17,3 +21,76 @@ func (h *Handler) Ping(c *fiber.Ctx) error {
...
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (h *Handler) GetPoster(c *fiber.Ctx) error {
+	chatID := c.Params("chat_id")
```
Comment (cringoleg) [inline internal/controller.go]: What chat?
Also move to const.

File: internal/core/db/db.go [hunk 13]
```diff
@@ -1,22 +1,31 @@
...
-	dsnRaw := "user=%s password=%s host=%s port=%s dbname=%s sslmode=disable"
+	gormlog "gorm.io/gorm/logger"
+)
 
-	dsn := fmt.Sprintf(dsnRaw,
-		dbConf.DB.User, dbConf.DB.Pass, dbConf.DB.Host, dbConf.DB.Port, dbConf.DB.Name)
+func NewSqlDB(ctx context.Context, dbUrl string) (*gorm.DB, *pgxpool.Pool, error) {
+	gormInstance, err := gorm.Open(postgres.Open(dbUrl), &gorm.Config{
+		Logger: gormlog.Default.LogMode(gormlog.Info),
+	})
```
Comment (cringoleg) [inline internal/core/db/db.go]: Break line for each argument passed to `gorm.Open`

**Review by cringoleg — CHANGES_REQUESTED** [review]
Check all other occurrences of `return err` and replace them with errors.Wrap() if possible.
Replace all static strings with constants if possible.

File: internal/storage.go [hunk 14]
```diff
@@ -1,11 +1,57 @@
...
+		Update("deleted_at", time.Now()).Error
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *Storage) UpdatePoster(p entity.Poster) error {
+	err := s.gorm.Updates(&p).Where("deleted_at IS NULL").Error
```
Comment (cringoleg) [inline internal/storage.go]: Replace with const?

File: internal/storage.go [hunk 15]
```diff
@@ -1,11 +1,57 @@
...
+	if err != nil {
+		return entity.Poster{}, err
+	}
+
+	return poster, nil
+}
+
+func (s *Storage) DeletePoster(pID string) error {
+	err := s.gorm.Table(posterTable).Where("id = ? AND deleted_at IS NULL", pID).
+		Update("deleted_at", time.Now()).Error
```
Comment (cringoleg) [inline internal/storage.go]: Replace with const?

**Review by cringoleg — CHANGES_REQUESTED** [review]

File: database/migrate.go (hunk 1 above)
Comment (yaroslavvasilenko) [reply database/migrate.go:18]: I need the standard interface, it just uses a different driver.

What works, we don't touch.

File: internal/storage.go (hunk 15 above)
Comment (yaroslavvasilenko) [reply internal/storage.go]: No. It is sql

**Pull request merged**